README.md
# wekan
Wekan auto-installation script

Tested on Ubuntu Server 14.04.4 LTS but it should work on any Debian based distribution.

This scripts automates the installation process for **[Wekan v0.10.1](https://github.com/wekan/wekan/releases/tag/v0.10.1)**

It automatically downloads and installs the following packages:

* make, gcc, g++, build-essential, libssl-dev (needed for wekan compilation)
* mongodb-server (NoSQL database)
* nvm (installs NodeJS v0.10.40)
* forever (runs NodeJS applications at the background forever)
* wekan v0.10.1

## Pre-Configuration
Before you start the installation, you can edit the script to change some parameters, such as:

* *NODE_VERSION* (default is v0.10.40 which is required for running Wekan v0.10.1. I suggest **NOT CHANGING IT** since it could break the service)
* *MONGO_URL* (default is mongodb://127.0.0.1:27017/wekan)
* *ROOT_URL* (default is http://127.0.0.1)
* *PORT* (defaults is set to **8080**)
* *MAIL_URL* (the format is 'smtp://user:pass@mailserver.examples.com:25/')
* *PIDFILE* (default is /var/run/Wekan.pid)
* *LOGFILE* (default is /var/run/Wekan.log)


## Installation
Run as root:

```sh
$ chmod +x autoinstall_wekan.sh
$ ./autoinstall_wekan.sh
```

As a plus, the script configures Wekan as a service which autostarts at boot.

After the installation is complete, you have the option to start the service or to reboot the system.


## Service
You can manually start the service as root by typing:

```sh
$ /etc/init.d/wekan start
```

## Access your installation

After the service is started you can access your fresh wekan installation by pointing your browser to:

```
http://<your-ip>:8080
```

The default port is set to **8080** in case you have another web server running at port 80.

## Post-Configuration

In case want to reconfigure the service to run at a different port after the you run the script, you must edit the service file using your favorite file editor, for ex.

```sh
$ nano /etc/init.d/wekan
```

Change any parameters you want, save and exit. After that you must update the autostart service script. To do so run as root:

```sh
$ update-rc.d -f wekan remove && update-rc.d wekan defaults
```

For the changes to take effect you should **restart the service**. I recommend **rebooting your system**, since the previous service could be still running.

## Wekan Virtual Appliance

I have prepared a ready to run, Ubuntu based (Ubuntu 14.04.4 Server - amd64) Virtual Machine with [Wekan](https://github.com/wekan/wekan) preinstalled. All you need is a VirtualBox installation. You can download the appliance directly from https://selalmaz.xyz/t/wekanmv-0-10-1 (~689MB) or from the releases page at https://github.com/anselal/wekan/releases

* username: wekan
* password: wekan


CHANGELOG.md

# v0.10.1

* Updated VM to Ubuntu Server 14.04.4 LTS
* Check if curl is installed and install it if not
* Use aptitude instead of apt-get if it exists
* Custom script that shows the ip address at the login prompt
* SSH service installed


autoinstall_wekan.sh
#!/bin/bash

if [[ $EUID -ne 0 ]]; then
   echo ""
   echo "This script must be run as root" 1>&2
   echo ""
   exit 1
fi

which aptitude > /dev/null
# Use aptitude if it exists
if [ $? -eq "0" ]; then
	aptitude update && aptitude install -y mongodb-server make g++ gcc build-essential libssl-dev
# else us apt-get
elif [ $? -ne "0" ]; then
	apt-get update && apt-get install -y mongodb-server make g++ gcc build-essential libssl-dev
fi

aptitude update && aptitude install -y mongodb-server make g++ gcc build-essential libssl-dev

# wget -qO- https://raw.githubusercontent.com/creationix/nvm/v0.30.0/install.sh | bash

# Check if curl exists
which curl > /dev/null
# If curl does not exist, install it
if [ $? -ne "0" ]; then
    echo "curl does not exist. I will install it for you..."
    apt-get install curl
fi

#curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.30.0/install.sh | bash
curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.30.2/install.sh | bash

source ~/.nvm/nvm.sh
source ~/.profile
source ~/.bashrc

# Make node, npm, available to all users
# n=$(which node);n=${n%/bin/node}; chmod -R 755 $n/bin; sudo cp -r $n/{bin,lib,share} /usr/local

NODE_VERSION="v0.10.40"

nvm install $NODE_VERSION
nvm use $NODE_VERSION
nvm alias default $NODE_VERSION

npm install -g npm
npm install forever -g

cd
wget https://github.com/wekan/wekan/releases/download/v0.10.1/wekan-0.10.1.tar.gz
rm -rf ~/bundle
tar xzvf wekan-0.10.1.tar.gz

cd bundle/programs/server && npm install

echo ""
echo "Installing wekan service..."
echo -ne '#####                     (33%)\r'
sleep 1
echo -ne '#############             (66%)\r'
sleep 1
echo -ne '#######################   (100%)\r'
echo -ne '\n'

# Cat text, but keep $ uninterpreted
cat > /etc/init.d/wekan <<'_EOF'
#!/bin/bash
#/etc/init.d/wekan

### BEGIN INIT INFO
# Provides:          wekan
# Required-Start:    $all
# Required-Stop:     $all
# Should-Start:      $mongodb
# Should-Stop:
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: Start daemon at boot time
# Description:       Enable service provided by daemon.
### END INIT INFO

_EOF

# Cat text, but interpret $
cat >> /etc/init.d/wekan <<_EOF
export NODE_PATH=/root/.nvm/$NODE_VERSION/bin
_EOF

# Cat text, but keep $ uninterpreted
cat >> /etc/init.d/wekan <<'_EOF'
export PATH=$PATH:$NODE_PATH:/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin
export MONGO_URL="mongodb://127.0.0.1:27017/wekan"
export ROOT_URL="http://127.0.0.1"
export PORT="8080"
export MAIL_URL='smtp://user:pass@mailserver.examples.com:25/'

NAME="Wekan"
APPLICATION_DIRECTORY=/root/bundle
APPLICATION_START=main.js
PIDFILE=/var/run/$NAME.pid
LOGFILE=/var/run/$NAME.log
FOREVER=$(which forever)

start() {
	echo "Starting $NAME"
	$NODE_PATH/forever --pidFile $PIDFILE --sourceDir=$APPLICATION_DIRECTORY -a -l $LOGFILE --minUptime 5000 --spinSleepTime 2000 start $APPLICATION_START &
	RETVAL=$?

}

stop() {
	if [ -f $PIDFILE ]; then
		echo "Shutting down $NAME"
		rm -f $PIDFILE
		RETVAL=$?
	else
		echo "$NAME is not running."
		RETVAL=0
	fi
}

restart() {
	echo "Restarting $NAME"
	stop
	start
}

status() {
	echo "Status for $NAME"
	$NODE_PATH/forever list
	RETVAL=$?
}


case "$1" in
  start)
	start
	;;
  stop)
	stop
	;;
  status)
	status
	;;
  restart)
	restart
	;;
  *)
	echo "Usage: /etc/init.d/wekan {start|stop|status|restart}"
	exit 1
	;;
esac

exit $RETVAL
_EOF

chmod +x /etc/init.d/wekan
update-rc.d -f wekan remove
update-rc.d wekan defaults

echo ""
echo "Installing custom scripts..."

# First step
cp /etc/issue /etc/issue-standard
echo -ne '#####                     (33%)\r'
sleep 1
# Second step
cat > /usr/local/bin/get-ip-address <<'_EOF'
#!/bin/bash
/sbin/ifconfig | grep "inet addr" | grep -v "127.0.0.1" | awk '{ print $2 }' | awk -F: '{ print $2 }'

_EOF
chmod +x /usr/local/bin/get-ip-address
echo -ne '#############             (66%)\r'
sleep 1
# Third step
cat > /etc/network/if-up.d/show-ip-address <<'_EOF'
#!/bin/bash
if [ "$METHOD" = loopback ]; then
    exit
fi

# Only run from ifup
if [ "$MODE" != start ]; then
    exit 0
fi

cp /etc/issue-standard /etc/issue
#IP="`ip route get 8.8.8.8 | awk '{ $NF; exit; }'`"
IP="IP ADDRESS: `/usr/local/bin/get-ip-address` (default port 8080)"
echo -e "\e[33m$IP \e[39m " >> /etc/issue
echo "" >> /etc/issue

_EOF
chmod +x /etc/network/if-up.d/show-ip-address
echo -ne '#######################   (100%)\r'
sleep 1
echo -ne '\n'


echo ""
echo "Installation completed."
menu() {
echo ""
echo "You can run the service manually or you can reboot your system."
echo "Please select one of the following:"
echo "1) Start the wekan service"
echo "2) Reboot your system"
echo "3) Exit"
echo -n "Choice: "
read choice

case "$choice" in
1)
exec /etc/init.d/wekan start
;;
2)
echo "Rebooting the system..."
sleep 3
exec reboot
;;
3)
echo "Exiting."
sleep 1
exit 0
;;
*)
menu
;;
esac
}
menu